objectherkenning_openbare_ruimte/on_edge/data_delivery_pipeline/components/data_delivery.py

```diff
@@ -0,0 +1,267 @@
+            self.settings["data_delivery_pipeline"]["project_version"],
+        ]
+        self.detection_metadata_header = [
+            "image_name",
+            "object_class",
+            "x_center",
+            "y_center",
+            "width",
+            "height",
+            "confidence",
+            "tracking_id",
+        ]
+
+    def run_pipeline(self):
+        """
+        Runs the data delivery pipeline:
+            - retrieves all the csvs and images that need to be delivered;
+            - delivers the data to Azure;
+            - deletes the delivered data.
+        """
+        logger.info(f"Running delivery pipeline on {self.detections_folder}..")
+        metadata_csv_file_paths = get_frame_metadata_csv_file_paths(
+            root_folder=self.detections_folder
+        )
+        logger.info(f"Number of CSVs to deliver: {len(metadata_csv_file_paths)}")
+        self._deliver_data(frame_metadata_file_paths=metadata_csv_file_paths)
+        self._delete_data(metadata_csv_file_paths=metadata_csv_file_paths)
+
+    def _deliver_data(self, frame_metadata_file_paths):
+        """
+        Using Azure IoT delivers the images and metadata to Azure.
+
+        Parameters
+        ----------
+        frame_metadata_file_paths
+            CSV files containing the metadata of the pictures,
+            it's used to keep track of which files need to be delivered.
+
+        """
+        iot_handler = IoTHandler(
+            hostname=self.settings["azure_iot"]["hostname"],
+            device_id=self.settings["azure_iot"]["device_id"],
+            shared_access_key=self.settings["azure_iot"]["shared_access_key"],
+        )
+        for frame_metadata_file_path in frame_metadata_file_paths:
+            (
+                csv_path,
+                relative_path,
+                detections_path,
+                path_only_filtered_rows,
+                file_path_only_filtered_rows,
+                file_path_detection_metadata,
+            ) = self._calculate_all_paths(
+                metadata_csv_file_path=frame_metadata_file_path
+            )
+
+            filtered_frame_metadata_rows = []
+            detection_metadata_rows = []
+
+            try:
+                with open(frame_metadata_file_path) as frame_metadata_file:
+                    reader = csv.reader(frame_metadata_file)
+                    header = next(reader)
+                    new_frame_metadata_header = (
+                        ["image_name"]
+                        + header
+                        + ["model_name", "model_version", "code_version"]
+                    )
+                    filtered_frame_metadata_rows.append(new_frame_metadata_header)
+                    detection_metadata_rows.append(self.detection_metadata_header)
+
+                    images_delivered = 0
+                    for row in reader:
+                        image_file_name = get_img_name_from_csv_row(csv_path, row)
```

Comment:
I think this is what needs to be changed to make it work with subfolders in the storage
file_path should be "detection_metadata or frame_metadata or images /Date/file_path"